src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,141 @@
+import re
+from collections import deque
 from typing import Optional, Type
 
-from ape.exceptions import ContractLogicError, TransactionError
+import ape
+from ape.exceptions import APINotImplementedError, ContractLogicError, TransactionError
 
 
 class RevertsContextManager:
-    def __init__(self, expected_message: Optional[str] = None):
+    def __init__(self, expected_message: Optional[str] = None, dev_message: Optional[str] = None):
+        if dev_message is not None and expected_message is not None:
+            raise ValueError("Cannot supply both 'expected_message' and 'dev_message'.")
+
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exception: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        Raises:
+            AssertionError: When the trace or source can not be retrieved, the trace does not end in
+            a REVERT instruction, or the dev message cannot be found or does not match the expected
+            dev message.
+        """
+        if exception.txn is None:
+            raise AssertionError("Could not fetch transaction information to check dev message.")
 
-        if not isinstance(exc_value, ContractLogicError):
+        receipt = exception.txn.provider.get_receipt(
+            exception.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
+        except ValueError as exc:
             raise AssertionError(
-                f"Transaction did not revert.\n"
-                f"However, an exception of type {type(exc_value)} occurred: {exc_value}"
-            ) from exc_value
+                f"Could not fetch contract at {receipt.receiver} to check dev message."
+            ) from exc
+
+        if contract.contract_type.pcmap is None:
+            raise AssertionError("Compiler does not support source code mapping.")
+
+        try:
+            trace = deque(exception.trace)
```

Comment:
Another Mypy question:

I've technically already established the pre-condition earlier on by bailing early if `exception.txn is None`

So here would I just use `# type: ignore`? Want to make sure that's the acceptable way to solve these issues with the type checking when it's complaining about situations that can't actually happen

cc @NotPeopling2day @unparalleled-js

Reply:
One way I prefer is save it is as a local variable first.

```python
txn = exception.txn
```

From a typing-checking perspective, the implementation of the calculated property could produce different results each time it was called, thus the type is not ever considered fully non-optional.

Reply:
And then do the type-checking on the local variable

Reply:
@unparalleled-js in this case it's not `txn`, but rather that the `trace` property would only return `None` if the `txn` happened to be `None` as well. So by checking the transaction first the prerequisite is already set that `trace` cannot be returning `None`. Would I just `# type: ignore` in that case?

(this applies whether [we change to `receipt`](https://github.com/ApeWorX/ape/pull/1084#discussion_r1010953444) or not)

Reply:
If you are not too worried, you can use `type: ignore` with an additional comment explaining it.
My aim is suggesting not to do that is preventing possible future bugs but complicating the implementation of `trace()` such that `None` is returned in other scenarios besides when the txn is `None`.

Reply:
Updated to specifically check `if exception.trace is None`